components/common-go/watch/file.go

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package watch
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/fsnotify/fsnotify"
+	"golang.org/x/xerrors"
+
+	"github.com/gitpod-io/gitpod/common-go/log"
+)
+
+type fileWatcher struct {
+	onChange func()
+
+	watcher *fsnotify.Watcher
+
+	hash string
+}
+
+func File(ctx context.Context, path string, onChange func()) error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return xerrors.Errorf("unexpected error creating file watcher: %w", err)
+	}
+
+	fw := &fileWatcher{
+		watcher:  watcher,
+		onChange: onChange,
+	}
+
+	// initial hash of the file
+	hash, err := hashConfig(path)
+	if err != nil {
+		return xerrors.Errorf("cannot get hash of file %v: %w", path, err)
+	}
+
+	watchDir, _ := filepath.Split(path)
+	err = watcher.Add(watchDir)
+	if err != nil {
+		watcher.Close()
+		return xerrors.Errorf("unexpected error watching file %v: %w", path, err)
+	}
+
+	log.Infof("starting watch of file %v", path)
+
+	fw.hash = hash
+
+	go func() {
+		defer func() {
+			log.WithError(err).Error("stopping file watch")
+
+			err = watcher.Close()
+			if err != nil {
+				log.WithError(err).Error("unexpected error closing file watcher")
+			}
+		}()
+
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok {
+					return
+				}
+
+				if !eventOpIs(event, fsnotify.Create) && !eventOpIs(event, fsnotify.Remove) {
```

Comment:
Should we skip [rename](https://github.com/fsnotify/fsnotify/blob/3eb7fa4b3213c056d4657a9776041e96e8fe73b6/fsnotify.go#L31) and [chmod](https://github.com/fsnotify/fsnotify/blob/3eb7fa4b3213c056d4657a9776041e96e8fe73b6/fsnotify.go#L32) as well?

Reply:
It does already. This skips everything except for `create` and `remove` ops.

Reply:
Oops, I did not see the `!`.

So, if we update the Kubernetes ConfigMap, the behaviour of the file is recreating the file (delete->create)?